lib/ASRouter.jsm

```diff
@@ -136,6 +142,38 @@ class _ASRouter {
     this.onMessage = this.onMessage.bind(this);
   }
 
+  _addASRouterPrefListener() {
+    if (this._prefs === null) {
+      this._prefs = Services.prefs.getBranch(ASROUTER_PREF_BRANCH);
+      this._prefs.addObserver("", this);
+    }
+  }
+
+  // Update provider endpoint and fetch new messages on pref change
+  async observe(aSubject, aTopic, aPrefName) {
+    if (!this.initialized) {
+      return;
+    }
+
+    await this.setState(prevState => {
+      const providers = [...prevState.providers];
+      this._updateProviderEndpointUrl(providers.find(p => p.endpointPref === aPrefName));
+      return {providers};
+    });
+
+    await this.loadMessagesFromAllProviders();
+  }
+
+  _updateProviderEndpointUrl(provider) {
```

Comment:
Because this is running in the constructor, I think this is running into some race conditions with our AS-specific prefs getting added later (i.e. it basically returns the default `""` and never gets updated)

In order to make this less bug prone, I would maybe checking the pref / setting `provider.url` at the point at which messages are fetched (i.e. in `_remoteLoader`). Fetches don't happen so often so I wouldn't worry about the overhead. You can still reset `.lastUpdated` and reload messages in the observer.

Reply:
Wanted to add some `broadcast` method to update opened tabs when pref is changed but getting the next message is [tied to the target](https://github.com/mozilla/activity-stream/blob/c97d8464ddefc9109c06b0137f7dea48434b7419/lib/ASRouter.jsm#L323) and doesn't make it very straight forward if you want to broadcast the same message to all message channel ports.
Does the feature make sense? Should it be a followup? 
Only noticed it because ASRouter admin reacts to pref changes and updates the message provider header.